apis/bigqueryconnection/v1alpha1/connection_reference.go

```diff
@@ -0,0 +1,156 @@
+// holds the GCP identifier for the KRM object.
+type BigQueryConnectionConnectionRef struct {
+	// A reference to an externally managed BigQueryConnectionConnection resource.
+	// Should be in the format `projects/<projectID>/locations/<location>/connections/<connectionID>`.
+	External string `json:"external,omitempty"`
+
+	// The `name` of a `BigQueryConnectionConnection` resource.
+	Name string `json:"name,omitempty"`
+	// The `namespace` of a `BigQueryConnectionConnection` resource.
+	Namespace string `json:"namespace,omitempty"`
+
+	parent string
+}
+
+func (r *BigQueryConnectionConnectionRef) Parent() (string, error) {
+	if r.parent != "" {
+		return r.parent, nil
+	}
+	if r.External != "" {
+		r.External = strings.TrimPrefix(r.External, "/")
+		tokens := strings.Split(r.External, "/")
+		if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "connections" {
+			return "", fmt.Errorf("format of BigQueryConnectionConnection external=%q was not known (use projects/<projectId>/locations/<location>/connections/<connectionID>)", r.External)
+		}
+		r.parent = "projects/" + tokens[1] + "/locations/" + tokens[3]
+		return r.parent, nil
+	}
+	return "", fmt.Errorf("BigQueryConnectionConnectionRef not normalized to External form or not created from `New()`")
+}
+
+// NormalizeToExternalForm provision the "External" value.
+// If the "External" comes from the ConfigConnector object, it has to acquire or reconcile with the GCP resource already.
+func (r *BigQueryConnectionConnectionRef) NormalizeToExternalForm(ctx context.Context, reader client.Reader) error {
+	if r.External != "" && r.Name != "" {
+		return fmt.Errorf("cannot specify both name and external on %s reference", BigQueryConnectionConnectionGVK.Kind)
+	}
+	if r.External != "" {
+		r.External = strings.TrimPrefix(r.External, "/")
+		tokens := strings.Split(r.External, "/")
+		if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "connections" {
+			return fmt.Errorf("format of BigQueryConnectionConnection external=%q was not known (use projects/<projectId>/locations/<location>/connections/<connectionID>)", r.External)
+		}
+		return nil
+	}
+	key := types.NamespacedName{Name: r.Name, Namespace: r.Namespace}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(BigQueryConnectionConnectionGVK)
+	if err := reader.Get(ctx, key, u); err != nil {
+		if apierrors.IsNotFound(err) {
+			return k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+		}
+		return fmt.Errorf("reading referenced %s %s: %w", BigQueryConnectionConnectionGVK, key, err)
+	}
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef == "" {
+		return fmt.Errorf("BigQueryConnectionConnection is not ready yet.")
```

Comment:
Nit (no need to fix here) - I think we want to start returning our special "not ready" error to trigger our special backoff-with-auto-watch etc

Reply:
Wait! what did i miss?